static/js/components/forms/CouponForm.js

```diff
@@ -0,0 +1,298 @@
+// @flow
+import React from "react"
+import Picky from "react-picky"
+import DayPickerInput from "react-day-picker/DayPickerInput"
+import { Formik, Field, Form, ErrorMessage } from "formik"
+import * as yup from "yup"
+
+import type { Company, Product } from "../../flow/ecommerceTypes"
+
+type CouponFormProps = {
+  onSubmit: Function,
+  toggleForm: Function,
+  toggleProduct: Function,
+  isPromo: boolean,
+  companies: Array<Company>,
+  products: Array<Product>,
+  selectProducts: Function,
+  selectedProducts: Array<Product>,
+  productType: string
+}
+
+const baseValidations = {
+  name: yup
+    .string()
+    .required("Coupon name is required")
+    .matches(/^\w+$/, "Only letters, numbers, and underscores allowed"),
+  products:        yup.array().min(1, "${min} or more products must be selected"),
+  activation_date: yup.date().required("Valid activation date required"),
+  expiration_date: yup.date().required("Valid expiration date required"),
+  amount:          yup
+    .number()
+    .required("Percentage discount is required")
+    .min(1, "Must be at least ${min}")
+    .max(100, "Must be at most ${max}")
+}
+
+const promoValidations = yup.object(baseValidations).shape({
+  max_redemptions: yup
+    .number()
+    .min(1, "Must be at least ${min}")
+    .required("Number required"),
+  coupon_code: yup
+    .string()
+    .required("Coupon code is required")
+    .matches(/^\w+$/, "Only letters, numbers, and underscores allowed")
+})
+
+const singleUseValidations = yup.object(baseValidations).shape({
+  num_coupon_codes: yup
+    .number()
+    .min(1, "Must be at least ${min}")
+    .required("Number required"),
+  payment_transaction: yup.string().required("Payment transaction is required"),
+  payment_type:        yup.string().required("Payment type is required")
+})
+
+export const CouponForm = ({
+  onSubmit,
+  toggleForm,
+  toggleProduct,
+  isPromo,
+  companies,
+  products,
+  productType,
+  selectProducts,
+  selectedProducts
+}: CouponFormProps) => (
+  <Formik
+    onSubmit={onSubmit}
+    validationSchema={isPromo ? promoValidations : singleUseValidations}
+    initialValues={{
+      coupon_type:         isPromo ? "promo" : "single-use",
+      product_type:        productType,
+      products:            [],
+      amount:              "",
+      name:                "",
+      num_coupon_codes:    1,
+      coupon_code:         "",
+      activation_date:     "",
+      expiration_dat:      "",
+      payment_type:        "",
+      payment_transaction: "",
+      company:             ""
+    }}
+    render={({
+      isSubmitting,
+      setFieldValue,
+      setFieldTouched,
+      errors,
+      touched
+    }) => (
+      <Form className="coupon-form">
+        <div className="flex">
+          <div>
+            <Field
+              type="radio"
+              name="coupon_type"
+              value="single-use"
+              onClick={toggleForm}
+              checked={!isPromo}
```

Comment:
It looks like you're storing the state of this in the page outside the form, which sort of breaks the idea of formik keeping form state internal to the form. I think you should be able to leave this to the defualt workings of formik and then conditionally swap out the inputs based on `values.coupon_type`